### PR TITLE
LC-466: create embedding stage using Jlama

### PR DESCRIPTION
**Pull request by Jozurf**
Note that there is an existing bug in Jlama which does not allow certain models to be parsed due to non-standardization of certain file formats when downloading a model.

### CODE REVIEW

File: lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    this.source = config.getString("source");
+    this.dest = config.hasPath("dest") ? config.getString("dest") : "embeddings";
+    this.workingMemoryType = config.hasPath("workingMemoryType") ? config.getString("workingMemoryType") : "";
+    this.workingQuantizationType = config.hasPath("workingQuantizationType") ? config.getString("workingQuantizationType") : "";
+  }
+
+  @Override
+  public void start() throws StageException {
+    try {
+      File localModelPath = SafeTensorSupport.maybeDownloadModel(pathToStoreModel, embeddingModel);
```
Comment (Jozurf) [inline lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java:77]: What do folks think about a "removeModelAfter" boolean stage property that removes the model folder using the path that it was stored at in the stop() method of the stage?

File: lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+ *    "F32" (32-bit floating point)
+ *    "F64" (64-bit floating point)
+ *    "I64" (Signed 64-bit integer)
+ *    "U64" (Unsigned 64-bit integer)
+ *    "Q4" (Quantized 4-bit integer)
+ *    "Q5" (Quantized 5-bit integer)
+ */
+
+public class JlamaEmbed extends Stage {
+  public final String embeddingModel;
```
Comment (rseitz) [inline lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java:57]: why are all these instance variables public?

File: lucille-plugins/lucille-jlama/src/test/java/com/kmwllc/lucille/jlama/stage/JlamaEmbedTest.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+import java.util.List;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.MockedStatic;
+
+public class JlamaEmbedTest {
+
+  private StageFactory factory = StageFactory.of(JlamaEmbed.class);
+  public Config defaultConfig;
```
Comment (rseitz) [inline lucille-plugins/lucille-jlama/src/test/java/com/kmwllc/lucille/jlama/stage/JlamaEmbedTest.java:34]: should be private

File: lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java (hunk 2 above)
Comment (Jozurf) [reply lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java]: overlooked these. Changed them to private in [27bd5f3](https://github.com/kmwtechnology/lucille/pull/201/commits/27bd5f36fae2a778926fe2c57c08df1b49bed931)

File: lucille-plugins/lucille-jlama/src/test/java/com/kmwllc/lucille/jlama/stage/JlamaEmbedTest.java (hunk 3 above)
Comment (Jozurf) [reply lucille-plugins/lucille-jlama/src/test/java/com/kmwllc/lucille/jlama/stage/JlamaEmbedTest.java]: changed to private in [27bd5f3](https://github.com/kmwtechnology/lucille/pull/201/commits/27bd5f36fae2a778926fe2c57c08df1b49bed931)

File: lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+ *    "I32" (Signed 32-bit integer)
+ *    "U32" (Unsigned 32-bit integer)
+ *    "F32" (32-bit floating point)
+ *    "F64" (64-bit floating point)
+ *    "I64" (Signed 64-bit integer)
+ *    "U64" (Unsigned 64-bit integer)
+ *    "Q4" (Quantized 4-bit integer)
+ *    "Q5" (Quantized 5-bit integer)
+ */
+
```
Comment (rseitz) [inline lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java:55]: extra space here

File: lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+  private final boolean deleteModelAfter;
+  private AbstractModel model;
+
+  private static final Logger log = LoggerFactory.getLogger(JlamaEmbed.class);
+
+  public JlamaEmbed(Config config) {
+    super(config, new StageSpec()
+        .withRequiredProperties("model", "source", "pathToStoreModel")
+        .withOptionalProperties("workingMemoryType", "workingQuantizationType", "dest", "deleteModelAfter"));
+    pathToStoreModel = config.getString("pathToStoreModel");
```
Comment (rseitz) [inline lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java:76]: minor: use "this." like the other vars here

File: lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    this.dest = config.hasPath("dest") ? config.getString("dest") : "embeddings";
+    this.workingMemoryType = config.hasPath("workingMemoryType") ? config.getString("workingMemoryType") : "";
+    this.workingQuantizationType = config.hasPath("workingQuantizationType") ? config.getString("workingQuantizationType") : "";
+    this.deleteModelAfter = config.hasPath("deleteModelAfter") ? config.getBoolean("deleteModelAfter") : false;
+  }
+
+  @Override
+  public void start() throws StageException {
+    try {
+      File localModelPath = SafeTensorSupport.maybeDownloadModel(pathToStoreModel, embeddingModel);
```
Comment (rseitz) [inline lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java:88]: what happens if we have more than one worker thread and two instances of this stage are starting around the same time, each one calling SafeTensorSupport.maybeDownloadModel()?

File: lucille-plugins/lucille-jlama/README.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
+## Prerequisites of the Lucille JLama plugin
+
+The Jlama plugin requires Java 20 and above. If you would like to use Lucille with a lower Java version without Jlama, follow this one time setup of Maven Toolchains. This setup ensures that a higher Java version is used only for the modules that require it.
+
+## Maven Toolchains Setup
+
+Note that the Jlama pom file is configured to automatically detect Java 21 SDKs. Change the version in the pom file if you are using Java 20 or higher than Java 21.
+
+1. Download your Java version from your favourite vendor.
```
Comment (Jozurf) [inline lucille-plugins/lucille-jlama/README.md:9]: change "favourite" with "favorite"

File: lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java (hunk 5 above)
Comment (Jozurf) [reply lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java]: changed in [a3112e3](https://github.com/kmwtechnology/lucille/pull/201/commits/a3112e36451580e2b2300c7d1cee915f0197e737)

File: lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java (hunk 6 above)
Comment (Jozurf) [reply lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java]: updated that line of code to be synchronized in [a3112e3](https://github.com/kmwtechnology/lucille/pull/201/commits/a3112e36451580e2b2300c7d1cee915f0197e737)

**Review by rseitz — APPROVED** [review]

**Review by spencersolomon6 — APPROVED** [review]

File: lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    this.workingMemoryType = config.hasPath("workingMemoryType") ? config.getString("workingMemoryType") : "";
+    this.workingQuantizationType = config.hasPath("workingQuantizationType") ? config.getString("workingQuantizationType") : "";
+    this.deleteModelAfter = config.hasPath("deleteModelAfter") ? config.getBoolean("deleteModelAfter") : false;
+  }
+
+  @Override
+  public void start() throws StageException {
+    try {
+      File localModelPath;
+      synchronized (JlamaEmbed.class) {
```
[the remark this answers is not in the document]
Comment (Jozurf) [reply lucille-plugins/lucille-jlama/src/main/java/com/kmwllc/lucille/jlama/stage/JlamaEmbed.java:91]: added code comment in [ac1a6ae](https://github.com/kmwtechnology/lucille/pull/201/commits/ac1a6ae840321b2f6182da2a059ab1222aeaffba)

**Pull request merged**